prow/cmd/peribolos/main.go

```diff
@@ -184,57 +184,88 @@ func configureOrgMembers(opt options, client configureOrgMembersClient, orgName
 		haveMembers.Insert(m.Login)
 	}
 
-	// Ensure desired state is sane
-	both := wantAdmins.Intersection(wantMembers)
-	if n := len(both); n > 0 {
-		return fmt.Errorf("%d users are both a member and admin: %v", n, both)
-	}
-
+	have := memberships{members: haveMembers, super: haveAdmins}
+	want := memberships{members: wantMembers, super: wantAdmins}
 	// Figure out who to remove
-	have := haveMembers.Union(haveAdmins)
-	want := wantMembers.Union(wantAdmins)
-	remove := have.Difference(want)
-
-	// Figure out who to invite and/or reconfigure
-	makeMember := wantMembers.Difference(haveMembers)
-	makeAdmin := wantAdmins.Difference(haveAdmins)
+	remove := have.all().Difference(want.all())
 
 	// Sanity check changes
-	if d := float64(len(remove)) / float64(len(have)); d > opt.maximumDelta {
+	if d := float64(len(remove)) / float64(len(have.all())); d > opt.maximumDelta {
 		return fmt.Errorf("cannot delete %d memberships or %.3f of %s (exceeds limit of %.3f)", len(remove), d, orgName, opt.maximumDelta)
 	}
 
-	var errs []error
+	teamMembers := sets.String{}
+	for _, team := range orgConfig.Teams {
+		teamMembers.Insert(team.Members...)
+		teamMembers.Insert(team.Maintainers...)
+	}
 
-	// March towards desired state
-	for u := range remove {
-		if err := client.RemoveOrgMembership(orgName, u); err != nil {
-			logrus.WithError(err).Warnf("RemoveOrgMembership(%s, %s) failed", orgName, u)
-			errs = append(errs, err)
+	if outside := teamMembers.Difference(want.all()); len(outside) > 0 {
+		return fmt.Errorf("all team members/maintainers must also be org members: %s", strings.Join(outside.List(), ", "))
+	}
+
+	adder := func(user string, super bool) error {
+		r := "member"
+		if super {
+			r = "admin"
+		}
+		om, err := client.UpdateOrgMembership(orgName, user, super)
+		if err != nil {
+			logrus.WithError(err).Warnf("UpdateOrgMembership(%s, %s, %t) failed", orgName, user, super)
+		} else if om.State == github.StatePending {
+			logrus.Infof("Invited %s to %s as a %s", user, orgName, r)
 		} else {
-			logrus.Infof("Removed %s from %s", u, orgName)
+			logrus.Infof("Set %s as a %s of %s", user, r, orgName)
 		}
+		return err
 	}
 
-	for u := range makeMember {
-		if m, err := client.UpdateOrgMembership(orgName, u, false); err != nil {
-			logrus.WithError(err).Warnf("UpdateOrgMembership(%s, %s, false) failed", orgName, u)
+	remover := func(user string) error {
```

Comment:
`adder` logs on success, should `remover`?

Reply:
Done